hudi-common/src/main/java/org/apache/hudi/common/table/read/HoodieUnmergedFileGroupRecordBuffer.java

```diff
@@ -0,0 +1,140 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.hudi.common.table.read;
+
+import org.apache.hudi.common.config.TypedProperties;
+import org.apache.hudi.common.engine.HoodieReaderContext;
+import org.apache.hudi.common.model.DeleteRecord;
+import org.apache.hudi.common.model.HoodieRecordMerger;
+import org.apache.hudi.common.table.log.KeySpec;
+import org.apache.hudi.common.table.log.block.HoodieDataBlock;
+import org.apache.hudi.common.table.log.block.HoodieDeleteBlock;
+import org.apache.hudi.common.util.Option;
+import org.apache.hudi.common.util.ValidationUtils;
+import org.apache.hudi.common.util.collection.ClosableIterator;
+import org.apache.hudi.common.util.collection.ExternalSpillableMap;
+import org.apache.hudi.common.util.collection.Pair;
+import org.apache.hudi.exception.HoodieException;
+
+import org.apache.avro.Schema;
+
+import java.io.IOException;
+import java.io.Serializable;
+import java.util.Arrays;
+import java.util.Iterator;
+import java.util.Map;
+
+public class HoodieUnmergedFileGroupRecordBuffer<T> extends HoodieBaseFileGroupRecordBuffer<T> {
+  // Used to order the records in the record map.
+  private Long recordIndex = 0L;
+
+  public HoodieUnmergedFileGroupRecordBuffer(HoodieReaderContext<T> readerContext,
+                                             Schema readerSchema,
+                                             Schema baseFileSchema,
+                                             Option<String> partitionNameOverrideOpt,
+                                             Option<String[]> partitionPathFieldOpt,
+                                             HoodieRecordMerger recordMerger,
+                                             TypedProperties payloadProps,
+                                             long maxMemorySizeInBytes,
+                                             String spillableMapBasePath,
+                                             ExternalSpillableMap.DiskMapType diskMapType,
+                                             boolean isBitCaskDiskMapCompressionEnabled) {
+    super(readerContext, readerSchema, baseFileSchema, partitionNameOverrideOpt, partitionPathFieldOpt,
+        recordMerger, payloadProps, maxMemorySizeInBytes, spillableMapBasePath, diskMapType, isBitCaskDiskMapCompressionEnabled);
+  }
+
+  @Override
+  protected boolean doHasNext() throws IOException {
+    ValidationUtils.checkState(baseFileIterator != null, "Base file iterator has not been set yet");
+
+    // Handle merging.
+    while (baseFileIterator.hasNext()) {
+      nextRecord = baseFileIterator.next();
+      return true;
+    }
+
+    // Handle records solely from log files.
+    if (logRecordIterator == null) {
+      logRecordIterator = records.values().iterator();
+    }
+
+    while (logRecordIterator.hasNext()) {
+      Pair<Option<T>, Map<String, Object>> nextRecordInfo = logRecordIterator.next();
+      if (nextRecordInfo.getLeft().isPresent()) {
```

Comment:
if this is a delete record, `nextRecordInfo.getLeft()` is empty. How to construct the delete record as T?

Reply:
Meanwhile, how to make the hasNext respect the key order?